--- a/plugins/cadabra/progs/init-cadabra.scm	2025-05-25 14:47:36.077037100 -0500
+++ b/plugins/cadabra/progs/init-cadabra.scm	2025-05-10 21:14:36.896027738 -0500
@@ -1,17 +1,28 @@
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;
-;; MODULE      : cadabra.scm
-;; DESCRIPTION : Initialize cadabra plugin
-;; COPYRIGHT   : (C) 1999  Joris van der Hoeven
+;; MODULE      : cadabra2.scm
+;; DESCRIPTION : Initialize cadabra2 plugin
+;; COPYRIGHT   : (C) 2025 Kasper Peeters
 ;;
-;; This software falls under the GNU general public license version 3 or later.
-;; It comes WITHOUT ANY WARRANTY WHATSOEVER. For details, see the file LICENSE
-;; in the root directory or <http://www.gnu.org/licenses/gpl-3.0.html>.
+;; This software falls under the GNU general public license and comes WITHOUT
+;; ANY WARRANTY WHATSOEVER. See the file $TEXMACS_PATH/LICENSE for details.
+;; If you don't have this file, write to the Free Software Foundation, Inc.,
+;; 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 ;;
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
+;; As in the python plugin, we need a way to make TeXmacs send
+;; multi-line input without stripping newlines.
+
+(define (cadabra-serialize lan t)
+  (with u (pre-serialize lan t)
+    (with s (texmacs->code (stree->tree u) "SourceCode")
+	  (string-append  s  "\n<EOF>\n"))))
+
 (plugin-configure cadabra
-  (:require (url-exists-in-path? "cadabra"))
-  (:launch "cadabra --texmacs")
-  (:session "Cadabra"))
+  (:require (url-exists-in-path? "cadabra2"))
+  (:launch "cadabra2 --texmacs")
+  (:serializer ,cadabra-serialize)
+  (:session "Cadabra2"))
+
